funsor/terms.py

```diff
@@ -20,34 +20,21 @@
 from funsor.six import getargspec, singledispatch
 
 
-@dispatch(object, object)
 def substitute(expr, subs):
-    return expr  # default: do nothing
-
-
-@substitute.register(object, (OrderedDict, dict))
-def substitute_with_tuple(expr, subs):
-    return substitute(expr, tuple(subs.items()))
-
-
-@substitute.register(tuple, tuple)
-def substitute_tuple(expr, subs):
-    return tuple(substitute(v, subs) for v in expr)
-
-
-@substitute.register(Domain, tuple)
-def substitute_domain(expr, subs):
-    return expr
-
-
-@substitute.register(frozenset, tuple)
-def substitute_frozenset(expr, subs):
-    return frozenset(substitute(v, subs) for v in expr)
+    if isinstance(subs, (dict, OrderedDict)):
+        subs = tuple(subs.items())
+    assert isinstance(subs, tuple)
 
+    @interpreter.interpretation(interpreter._INTERPRETATION)  # use base
+    def subs_interpreter(cls, *args):
+        expr = cls(*args)
+        fresh_subs = tuple((k, v) for k, v in subs if k in expr.fresh)
+        if fresh_subs:
+            expr = interpreter.debug_logged(expr.eager_subs)(fresh_subs)
+        return expr
```

Comment:
I forget the `@interpreter.interpretation(-)` semantics. Does this first call `expr.eager_subs` and then pass the result for further reinterpretation by the base `interpreter._INTERPRETATION` captured above? If so then :+1:

Reply:
The base interpretation is invoked in the line above (`expr = cls(*args)`).